packages/fxa-event-broker/lib/serviceNotifications.ts

```diff
@@ -0,0 +1,124 @@
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import { Logger } from 'mozlog';
+import * as joi from 'typesafe-joi';
+
+// Event strings
+export const LOGIN_EVENT = 'login';
+export const SUBSCRIPTION_UPDATE_EVENT = 'subscription:update';
+export const DELETE_EVENT = 'delete';
+
+// Message schemas
+const BASE_MESSAGE_SCHEMA = joi
+  .object()
+  .keys({
+    event: joi.string().required()
+  })
+  .unknown(true)
+  .required();
+
+const LOGIN_SCHEMA = joi
+  .object()
+  .keys({
+    clientId: joi.string().optional(),
+    deviceCount: joi
+      .number()
+      .integer()
+      .required(),
+    email: joi.string().required(),
+    event: joi
+      .string()
+      .valid(LOGIN_EVENT)
+      .required(),
+    service: joi.string().required(),
+    uid: joi.string().required(),
+    userAgent: joi.string().required()
+  })
+  .unknown(true)
+  .required();
+
+const SUBSCRIPTION_UPDATE_SCHEMA = joi
+  .object()
+  .keys({
+    event: joi
+      .string()
+      .valid(SUBSCRIPTION_UPDATE_EVENT)
+      .required(),
+    isActive: joi.bool().required(),
+    productCapabilities: joi
+      .array()
+      .items(joi.string())
+      .required(),
+    productName: joi.string().required(),
+    subscriptionId: joi.string().required(),
+    uid: joi.string().required()
+  })
+  .unknown(true)
+  .required();
+
+const DELETE_SCHEMA = joi
+  .object()
+  .keys({
+    event: joi
+      .string()
+      .valid(DELETE_EVENT)
+      .required(),
+    uid: joi.string().required()
+  })
+  .unknown(true)
+  .required();
+
+export type loginSchema = joi.Literal<typeof LOGIN_SCHEMA>;
+export type subscriptionUpdateSchema = joi.Literal<typeof SUBSCRIPTION_UPDATE_SCHEMA>;
+export type deleteSchema = joi.Literal<typeof DELETE_SCHEMA>;
+
+export type ServiceNotification = loginSchema | subscriptionUpdateSchema | deleteSchema | undefined;
+
+interface SchemaTable {
+  [key: string]: joi.ObjectSchema<any>;
+}
+
+// Event Schema Map
+const eventSchemas = {
+  [LOGIN_EVENT]: LOGIN_SCHEMA,
+  [SUBSCRIPTION_UPDATE_EVENT]: SUBSCRIPTION_UPDATE_SCHEMA,
+  [DELETE_EVENT]: DELETE_SCHEMA
+};
+
+/**
+ * Attempt to validate and return one of a map of schemas based on the
+ * messages event type matching a schema map key.
+ *
+ * Note: Due to using `any` in the `SchemaTable` value, this is not actually
+ * type-safe.
+ *
+ * @param schemaTable
+ * @param rawMessage
+ */
+function multiSchemaAttempt(
+  schemaTable: SchemaTable,
+  rawMessage: object
+): ServiceNotification | undefined {
+  const message = joi.attempt(rawMessage, BASE_MESSAGE_SCHEMA);
+  if (schemaTable[message.event]) {
+    return joi.attempt(message, schemaTable[message.event]);
+  }
+}
+
+export const ServiceNotification = {
+  from(logger: Logger, rawMessage: object): ServiceNotification {
+    try {
+      const validMessage = multiSchemaAttempt(eventSchemas, rawMessage);
+      if (!validMessage) {
+        return undefined;
```

Comment:
no need to explicitly `return undefined` (unless this is TS magic at work)

Reply:
This is being used for Typescripts type refinement, to prove under the block that validMessage is no longer undefined. However I can drop a few lines of code by inverting the condition and letting the implicit undefined return value occur.